src/main/java/com/celements/cleverreach/DefaultFailNotificationHandler.java

```diff
@@ -0,0 +1,128 @@
+import com.celements.mailsender.IMailSenderRole;
+import com.celements.model.access.IModelAccessFacade;
+import com.celements.model.access.exception.DocumentNotExistsException;
+import com.celements.model.context.ModelContext;
+import com.celements.model.object.xwiki.XWikiObjectFetcher;
+import com.celements.web.classcollections.OldCoreClasses;
+import com.celements.web.classes.FormMailClass;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+
+@Component
+public class DefaultFailNotificationHandler implements FailNotificationHandlerRole {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(
+      DefaultFailNotificationHandler.class);
+
+  @Requirement
+  private ModelContext context;
+
+  @Requirement
+  private IModelAccessFacade modelAccess;
+
+  @Requirement
+  private IMailSenderRole mailSender;
+
+  @Requirement("celements.oldCoreClasses")
+  private IClassCollectionRole oldCoreClasses;
+
+  @Override
+  public void send(String msg, Exception excp) {
+    LOGGER.error(msg, excp);
+    try {
+      XWikiDocument configDoc = modelAccess.getDocument(getConfigDocRef());
+      List<BaseObject> receivers = XWikiObjectFetcher.on(configDoc).filter(
+          getReceiverEmailClassRef()).list();
+      String content = getMailingContent(msg, excp);
+      Optional<String> fromMail = getFromMail(configDoc);
+      if (fromMail.isPresent()) {
+        for (BaseObject receiver : receivers) {
+          if (1 == receiver.getIntValue("is_active")) {
+            mailSender.sendMail(fromMail.get(), null, receiver.getStringValue("email"), null,
+                null, "TAGESAGENDA UPDATE FAILED!", content, content, null, null);
+          }
+        }
+      } else {
+        LOGGER.error("Missing 'from' mail configuration for 'Tagesagenda update failed' "
+            + "notification");
+      }
+    } catch (DocumentNotExistsException dnee) {
+      LOGGER.error("Unable to read failed notification configuration. Doc does not exist.", dnee);
+    }
+  }
+
+  String getMailingContent(String msg, Exception excp) {
+    StringWriter content = new StringWriter();
+    content.append("<h2>")
+        .append(excp.getMessage())
+        .append("</h2><div>")
+        .append(msg).append("</div><hr /><pre>");
+    if ((excp instanceof CleverReachRequestFailedException)
+        && (((CleverReachRequestFailedException) excp).getResponse() != null)) {
+      Response resp = ((CleverReachRequestFailedException) excp).getResponse();
+      content.append("Status Code: ").append(Integer.toString(resp.getStatus())).append("\n");
+      String respHeaders = resp.getStringHeaders().entrySet().stream().collect(
+          Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().stream().collect(Collectors
+              .joining(" | ")))).entrySet().stream().map(entry -> entry.getKey() + " = " + entry
+                  .getValue() + "\n").collect(Collectors.joining());
```

Comment:
can be simplified to one map and collect statement.
```suggestion
      String respHeaders = resp.getStringHeaders().entrySet().stream()
          .map(entry -> entry.getKey() + " = " + entry.getValue().stream()
              .collect(Collectors.joining(" | ")) + "\n")
          .collect(Collectors.joining());
```